hazelcast/src/main/java/com/hazelcast/map/impl/operation/SynchronizeIndexesForPartitionTask.java

```diff
@@ -0,0 +1,112 @@
+import com.hazelcast.spi.impl.PartitionSpecificRunnable;
+import com.hazelcast.spi.serialization.SerializationService;
+import com.hazelcast.util.Clock;
+
+import java.util.Iterator;
+import java.util.List;
+
+// RU_COMPAT_38
+
+/**
+ * A task that is sent to all local partitions in order to add the indexes that have been specified in the task.
+ * Each task is executed for a separate partition.
+ *
+ * It is a compatibility mode between 3.8 and 3.9 nodes. It is only necessary due to race between
+ * PostJoinMapOperations and MapReplicationOperations, so the map data may arrive before the index definitions.
+ *
+ * This task is responsible to put the index instances back in sync with the definitions.
+ *
+ * @see com.hazelcast.map.impl.MapIndexSynchronizer
+ * @see com.hazelcast.map.impl.operation.PostJoinMapOperation
+ * @see com.hazelcast.map.impl.operation.MapReplicationStateHolder
+ */
+public class SynchronizeIndexesForPartitionTask implements PartitionSpecificRunnable {
+
+    private final int partitionId;
+    private final List<MapIndexInfo> mapIndexInfos;
+
+    private final MapService mapService;
+    private final SerializationService serializationService;
+    private final InternalPartitionService partitionService;
+
+    public SynchronizeIndexesForPartitionTask(int partitionId, List<MapIndexInfo> mapIndexInfos, MapService mapService,
+                                              SerializationService serializationService,
+                                              InternalPartitionService partitionService) {
+        this.partitionId = partitionId;
+        this.mapIndexInfos = mapIndexInfos;
+        this.mapService = mapService;
+        this.serializationService = serializationService;
+        this.partitionService = partitionService;
+    }
+
+    @Override
+    public void run() {
+        InternalPartition partition = partitionService.getPartition(partitionId, false);
+        if (!partition.isLocal() || partition.isMigrating()) {
+            return;
+        }
+
+        MapServiceContext mapServiceContext = mapService.getMapServiceContext();
+        long now = Clock.currentTimeMillis();
+
+        for (MapIndexInfo mapIndexInfo : mapIndexInfos) {
+            MapContainer mapContainer = mapServiceContext.getMapContainer(mapIndexInfo.getMapName());
+            Indexes indexes = mapContainer.getIndexes(partitionId);
+
+            // recreate missing indexes and
+            for (IndexInfo indexInfo : mapIndexInfo.getLsIndexes()) {
+                Index index = indexes.getIndex(indexInfo.getAttributeName());
+                if (index == null) {
+                    index = indexes.addOrGetIndex(indexInfo.getAttributeName(), indexInfo.isOrdered());
+                    RecordStore recordStore = mapServiceContext.getRecordStore(getPartitionId(), mapIndexInfo.getMapName());
+                    final Iterator<Record> iterator = recordStore.iterator(now, false);
+                    while (iterator.hasNext()) {
```

Comment:
I wasn't assuming multiple missing indexes, but you are right that iterating over RS is much more heavyweight, so I added the optimization anyway. I can't however modify the original list since it's reused locally for all partitions, for smaller memory overhead.